snapcraft/plugins/ruby.py

```diff
@@ -0,0 +1,142 @@
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The ruby plugin is useful for ruby based parts.
+The plugin uses gem to install dependencies from a `Gemfile`.
+This plugin uses the common plugin keywords as well as those for "sources".
+For more information check the 'plugins' topic for the former and the
+'sources' topic for the latter.
+Additionally, this plugin uses the following plugin-specific keywords:
+    - gems:
+      (list)
+      A list of gems to install.
+    - ruby-version:
+      (string)
+      The version of ruby you want this snap to run.
+"""
+import os
+import re
+import platform
+
+from snapcraft import BasePlugin
+from snapcraft.internal.errors import SnapcraftEnvironmentError
+from snapcraft.sources import Tar
+
+
+class RubyPlugin(BasePlugin):
+
+    @classmethod
+    def schema(cls):
+        schema = super().schema()
+
+        schema['properties']['ruby-version'] = {
+            'type': 'string',
+            'default': '2.3.1'
+        }
+        schema['properties']['gems'] = {
+            'type': 'array',
+            'minitems': 1,
+            'uniqueItems': True,
+            'items': {
+                'type': 'string'
+            },
+            'default': []
+        }
+        return schema
+
+    @classmethod
+    def get_pull_properties(cls):
+        # Inform Snapcraft of the properties associated with pulling. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return ['ruby-version', 'gems']
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+
+        self._ruby_version = self.options.ruby_version
+        self._ruby_part_dir = os.path.join(self.partdir, 'ruby')
+        self._ruby_download_url = \
+            'https://cache.ruby-lang.org/pub/ruby/ruby-{}.tar.gz'.format(
+                self._ruby_version)
+        self._ruby_tar = Tar(self._ruby_download_url, self._ruby_part_dir)
+        self._gems = self.options.gems or []
+        self._install_bundler = False
+        self._ruby_version_dir = None
+
+        self.build_packages.extend(['gcc', 'g++', 'make', 'zlib1g-dev',
+                                    'libssl-dev', 'libreadline-dev'])
+
+        version_map = [('2.%d.[0-9]' % i, '2.%d.0' % i) for i in range(5)]
+
+        for version_regex, version_dir in version_map:
+            if re.compile(version_regex).match(self._ruby_version):
+                self._ruby_version_dir = version_dir
+                break
+        # Raise exception when version_dir not found
+        if not self._ruby_version_dir:
+            raise SnapcraftEnvironmentError(
+                'Ruby version dir not in version_map.')
+
+    def pull(self):
+        super().pull()
+        os.makedirs(self._ruby_part_dir, exist_ok=True)
+        self._ruby_tar.download()
+        self._ruby_install(builddir=self._ruby_part_dir)
```

Comment:
I suggest not worrying about this for now, there are some more core issues here. We can fix it on iteration.